include/ginkgo/core/log/profiler_hook.hpp

```diff
@@ -0,0 +1,246 @@
+                                    const bool& set_finalized) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* solution, const uint8& stopping_id,
+        const bool& set_finalized, const array<stopping_status>* status,
+        const bool& one_changed, const bool& all_stopped) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* implicit_sq_resnorm, const LinOp* solution,
+        const uint8& stopping_id, const bool& set_finalized,
+        const array<stopping_status>* status, const bool& one_changed,
+        const bool& all_stopped) const override;
+
+    bool needs_propagation() const override;
+
+    /**
+     * Sets the name for an object to be profiled. Every instance of that object
+     * in the profile will be replaced by the name instead of its runtime type.
+     * @param obj  the object
+     * @param name  its name
+     */
+    void set_object_name(const PolymorphicObject* obj, std::string name);
+
+    /**
+     * Should the events call executor->synchronize on operations and
+     * copy/allocation? This leads to a certain overhead, but makes the
+     * execution timeline of kernels synchronous.
+     */
+    void set_synchronization(bool synchronize);
+
+    /** The Ginkgo yellow background color as packed 32 bit ARGB value. */
+    constexpr static uint32 color_yellow_argb = 0xFFFFCB05U;
+
+    /**
+     * Creates a logger annotating Ginkgo events with TAU ranges via PerfStubs.
+     * Since TAU requires global initialization and finalization, this will only
+     * create a single global instance that is destroyed at program exit.
+     * @param initialize  Should we call TAU's initialization and finalization
+     *                    functions, or does the application take care of it?
+     */
+    static std::shared_ptr<ProfilerHook> create_tau(bool initialize = true);
+
+    /**
+     * Creates a logger annotating Ginkgo events with NVTX ranges for CUDA.
+     * @param color_argb  The color of the NVTX ranges in the NSight Systems
+     *                    output. It has to be a 32 bit packed ARGB value.
+     */
+    static std::shared_ptr<ProfilerHook> create_nvtx(
+        uint32 color_argb = color_yellow_argb);
+
+    /**
+     * Creates a logger annotating Ginkgo events with ROCTX ranges for ROCm.
+     */
+    static std::shared_ptr<ProfilerHook> create_roctx();
+
+    /**
+     * Creates a logger annotating Ginkgo events with the most suitable backend
+     * for the given executor: NVTX for NSight Systems in CUDA, ROCTX for
+     * rocprof in ROCm, TAU for everything else.
+     */
+    static std::shared_ptr<ProfilerHook> create_for_executor(
```

Comment:
It isn't really using any properties of the executor, except for its type, which is why I found `for` more descriptive, but I am happy to be convinced otherwise 😄

Reply:
For me `from` seems more natural, but I don't want to die on that hill.